File: dataStructures/stackQeueDeque/deque.js
// Deque implemented with JS Array, using factory function

const createDeque = () => ({
  container: [],
  size: function() {
    return !this.container.length
  },
  isEmpty: function() {
    return this.container.length
  },
  addRear: function(item) {
    this.container.push(item)
  },
  addFront: function(item) {
    this.container.unshift(item)
  },
  removeRear: function() {
    return this.container.pop()
  },
  removeFront: function() {
    return this.container.shift()
  },
  serialize: function() {
    return this.container.join("-")
  }
})

const dq = createDeque()
console.log(dq.size()) // 0
dq.addFront("hello")
dq.addRear("world")
console.log(dq.serialize()) // "hello-world"
console.log(dq.isEmpty()) // false
console.log(dq.removeFront()) // "hello"
console.log(dq.removeRear()) // "world"

File: dataStructures/stackQeueDeque/python/deque.py
class Deque(object):
    def __init__(self):
        self.container = []

    def __len__(self):
        return len(self.container)

    def __str__(self):
        return "-".join([str(item) for item in self.container])

    def add_front(self, item):
        self.container.insert(0, item)

    def add_rear(self, item):
        self.container.append(item)

    def remove_front(self):
        return self.container.pop(0)

    def remove_rear(self):
        return self.container.pop()

    def is_empty(self):
        return not len(self.container)

if __name__ == "__main__":
    d = Deque()
    print(d.is_empty()) # True
    d.add_rear(4)
    d.add_rear("dog")
    d.add_front("cat")
    d.add_front(True)
    print(len(d)) # 4
    print(d.is_empty()) # False
    print(d.remove_front()) # True
    print(d.remove_rear()) # "dog"
    print(d) # "cat-4"
